ntoskrnl/io/iomgr/error.c

```diff
@@ -669,21 +677,29 @@ IoRaiseHardError(IN PIRP Irp,
         return;
     }
 
-    /* Setup an APC */
-    ErrorApc = ExAllocatePoolWithTag(NonPagedPool,
-                                     sizeof(KAPC),
-                                     TAG_APC);
+    // TODO: In case we were called in the context of a paging I/O or for
+    // a synchronous operation, that happens with APCs disabled, queue the
+    // hard-error call for later processing (see also IofCompleteRequest).
+
+    /* Setup an APC and queue it to deal with the error (see OSR documentation) */
+    ErrorApc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*ErrorApc), TAG_APC);
+    if (!ErrorApc)
+    {
+        /* Fail */
+        IoCompleteRequest(Irp, IO_DISK_INCREMENT);
+        return;
+    }
+
     KeInitializeApc(ErrorApc,
                     &Thread->Tcb,
                     Irp->ApcEnvironment,
-                    NULL,
                     IopFreeApc,
+                    NULL,
```

Comment:
(For information the allocated APC won't leak even if there is no RundownRoutine, because when such a function is absent, the kernel will then proceed to free the allocated APC:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ps/kill.c;hb=7790670dd80c5d4cf3c4b68f8327fe97126c6229#l880
)